### PR TITLE
Implemented User filters

### PR DESCRIPTION
**Pull request by nkartyshov**

### CODE REVIEW

File: adguard_cb/src/main/java/com/adguard/android/contentblocker/UserFilterActivity.java [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+        public UserFilterAdapter(Activity activity) {
+            this.preferencesService = ServiceLocator.getInstance(activity).getPreferencesService();
+            this.filterService = ServiceLocator.getInstance(activity).getFilterService();
+            this.activity = activity;
+
+            this.ruleList.addAll(preferencesService.getUserRules());
+        }
+
+        public void addRule(String rule) {
+            preferencesService.addUserRuleItem(rule);
```
Comment (ameshkov) [inline adguard_cb/src/main/java/com/adguard/android/contentblocker/UserFilterActivity.java]: At least trim the rule text

File: adguard_cb/src/main/java/com/adguard/android/contentblocker/UserFilterActivity.java [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+
+            new ApplyAndRefreshTask(filterService, activity).execute();
+
+            refreshAdapter();
+        }
+
+        public void removeRule(String rule) {
+            preferencesService.removeUserRuleItem(rule);
```
Comment (ameshkov) [inline adguard_cb/src/main/java/com/adguard/android/contentblocker/UserFilterActivity.java]: Same thing, you need to trim it.

File: adguard_cb/src/main/java/com/adguard/android/contentblocker/UserFilterActivity.java [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+import java.util.ArrayList;
+import java.util.List;
+
+public class UserFilterActivity extends AppCompatActivity implements View.OnClickListener {
+
+    private UserFilterAdapter userFilterAdapter;
+
+    @Override
+    protected void onCreate(@Nullable Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
```
Comment (ameshkov) [inline adguard_cb/src/main/java/com/adguard/android/contentblocker/UserFilterActivity.java]: Here is what's missing:

1. A placeholder when the user filter is empty.
2. Import/Export features
3. Clear button

Comment (nkartyshov) [inline adguard_cb/src/main/java/com/adguard/android/contentblocker/UserFilterActivity.java]: Fixed.

File: adguard_cb/src/main/java/com/adguard/android/service/FilterServiceImpl.java [hunk 4]
```diff
@@ -254,13 +254,19 @@ public boolean isSpywareFilterEnabled() {
 
     @Override
     public void applyNewSettings() {
-        // TODO fix this crutch
         setShowUsefulAds(preferencesService.isShowUsefulAds());
-        List<String> allEnabledRules = getAllEnabledRules(true);
-        cachedFilterRuleCount = allEnabledRules.size();
+
+        List<String> rules = getAllEnabledRules(true);
+        Set<String> userRules = getUserRules();
```
Comment (ameshkov) [inline adguard_cb/src/main/java/com/adguard/android/service/FilterServiceImpl.java]: Ignore non-ASCII rules here. They might be causing issues with Samsung browser.

Comment (ameshkov) [inline adguard_cb/src/main/java/com/adguard/android/service/FilterServiceImpl.java]: Also, trim every rule and check for control symbols.

Comment (nkartyshov) [inline adguard_cb/src/main/java/com/adguard/android/service/FilterServiceImpl.java]: Fixed

File: adguard_cb/src/main/java/com/adguard/android/service/FilterServiceImpl.java [hunk 5]
```diff
@@ -259,7 +265,19 @@ public void applyNewSettings() {
         List<String> rules = getAllEnabledRules(true);
         Set<String> userRules = getUserRules();
         if (!userRules.isEmpty()) {
-            rules.addAll(userRules);
+            for (String userRule : userRules) {
+                userRule = StringUtils.trim(userRule);
+                
+                if (StringUtils.isNotBlank(userRule) &&
```
Comment (ameshkov) [inline adguard_cb/src/main/java/com/adguard/android/service/FilterServiceImpl.java]: Wouldn't it be better to have a "validateRuleText" method instead.

Also, it'd be better to validate main filter rules as well.

Comment (nkartyshov) [inline adguard_cb/src/main/java/com/adguard/android/service/FilterServiceImpl.java]: Fixed

**Pull request merged**